integrate/simpsons_test.go

```diff
@@ -0,0 +1,82 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"math"
+	"testing"
+
+	"gonum.org/v1/gonum/floats"
+)
+
+func TestSimpsons(t *testing.T) {
+	const N = 1e6
+	x := floats.Span(make([]float64, N), 0, 1)
+	for i, test := range []struct {
+		x    []float64
+		f    func(x float64) float64
+		want float64
+	}{
+		{
+			x:    x,
+			f:    func(x float64) float64 { return x },
+			want: 0.5,
+		},
+		{
+			x:    floats.Span(make([]float64, N), -1, 1),
+			f:    func(x float64) float64 { return x },
+			want: 0,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return x + 10 },
+			want: 10.5,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return 3*x*x + 10 },
+			want: 11,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return math.Exp(x) },
+			want: 1.7182818284591876,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, math.Pi),
+			f:    func(x float64) float64 { return math.Cos(x) },
+			want: 0,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2*math.Pi),
+			f:    func(x float64) float64 { return math.Cos(x) },
+			want: 0,
+		},
+		{
+			x:    floats.Span(make([]float64, N*10), 0, math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 2,
+		},
+		{
+			x:    floats.Span(make([]float64, N*10), 0, 0.5*math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 1,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2*math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 0,
+		},
+	} {
+		y := make([]float64, len(test.x))
+		for i, v := range test.x {
+			y[i] = test.f(v)
+		}
+		v := Simpsons(test.x, y)
+		if !floats.EqualWithinAbs(v, test.want, 1e-4) {
```

Comment:
The same test for `Trapezoidal` uses `1e-12`. Why such a loose tolerance here? Maybe it's the inaccuracy I suspect above? This does not seem to accurately integrate even a linear function.

Reply:
Simpson's rule does not yield exact values for linear functions. We are making approximations with quadratic polynomials (parabolas). To describe f as points on a plane, this model best fits when the formed path/trajectory is curved and smooth (e.g. has arcs and not ridges, is quadratic and not linear).

I feel num:integrate models should be tested against the same corpus of functions. That being said, Simpson's rule will likely yield greater error than methods of numerical integration like Trapezium and Gaussian Quadrature.

If we were fuzzing these functions under test, I'd argue tolerance should be adjusted to appropriately fit each model, for example.

Reply:
If the formula can integrate exactly quadratic functions, then it can also integrate exactly linear and constant functions. How could it deal with the linear and constant terms otherwise? Linear and constant functions _are_ smooth.

For this reason, I agree that each test case should have it's own tolerance. Then we could set very stringent tolerances for functions that should be integrated exactly.

However, the reason for needing `1e-4` here is much simpler: there are several bugs in the algorithm. I'll comment on corresponding lines. When those are fixed, then with N=9 and tolerance `1e-12` only test cases 4, 6, 7, 8 fail. With N=100 only 6, 7, 8 fail. And with N=1000 all succeed.

Reply:
Yes, but not all functions we are interested in attempting to integrate are smooth. Please keep in mind those coefficients are not 'arbitrary constants'. 1) I feel one should never modify functions according to test cases. 2) By changing this implementation in that spirit I am concerned we are at risk of fitting an already extended Simpson's rule implementation to a group of equations.

Reply:
> Yes, but not all functions we are interested in attempting to integrate are smooth.

My comment said something different. Please read the discussion again. But now that you bring this up, I think we can say that all functions we can hope to integrate accurately with these simple methods have to be smooth and well-behaved. Simpson's rule and other similar methods are based on interpolating the sample points by polynomials, therefore they cannot approximate the underlying function well if it is not smooth or is even discontinuous. If there is such a point, then one should split the integration interval.

I don't understand the rest of your comment. I'm not saying you should change the weights, I'm saying that you should fix the bugs that I pointed out.